### PR TITLE
feat(carla_autoware): add interface to easily use CARLA with Autoware

### PR DESCRIPTION
**Pull request by mraditya01**
## Description

The interface between CARLA and Autoware aims to bridge both systems.
Required this launch PR: https://github.com/autowarefoundation/autoware_launch/pull/924

## Related links

- Main Repositories with instruction (https://github.com/mraditya01/autoware.universe/tree/updated_version/simulator/carla_autoware)
- CARLA Lanelet2 Maps (https://bitbucket.org/carla-simulator/autoware-contents/src/master/maps/)
- Update to e2e launch files (https://github.com/mraditya01/autoware_launch/tree/updated_version)
- Issues related to CARLA Autoware Bridge
> -  https://github.com/autowarefoundation/autoware.universe/issues/6487
> -  https://github.com/autowarefoundation/autoware.universe/issues/5701

## Tests performed

Tested on Town01 map from [Link](https://bitbucket.org/carla-simulator/autoware-contents/src/master/).

- **Environment Setup:** 
CARLA 0.9.15
Autoware.universe Main Branch
ROS2 Humble
Python 3.10
Ubuntu 22.04

- **Testing:** 
1. Tested the interface on the Town01 map provided by the CARLA simulator.
2. Verified the accuracy of vehicle localization within the simulated environment.
3. Checked that the vehicle's estimated position matches the ground truth position provided by CARLA.
4. Verified that the localization module in Autoware correctly interprets data received from CARLA.
5. Tested the perception module's ability to detect obstacle within the scene.
6. Tested vehicle control under various scenarios, including lane following, obstacle stop, and intersection.

- **Issues Related**

1. Tested velocity limit issue:

Confirmed the functionality of limiting vehicle velocity to 30 km/h by providing input via Autoware.
Verified that the vehicle's velocity did not exceed the specified limit during the test.

2. Tested vehicle stopped issue on particular path:

Replicated the reported issue by recreating the same path on the same map used in the initial observation.
Monitored the vehicle's behavior and trajectory during the test.
Confirmed that no issues occurred, indicating that the reported problem was not reproducible or no longer exists after further testing and investigation.

Video Working Example:
[Video](https://drive.google.com/file/d/1-WHcsY_Dou6rnsZJcN2v8DD-i54S6gNa/view?usp=sharing)

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: simulator/carla_autoware/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# CARLA_Autoware
```
Comment (maxime-clem) [inline simulator/carla_autoware/README.md:1]:
```suggestion
# CARLA Autoware
```

File: simulator/carla_autoware/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+# ROS2/Autoware.universe bridge for CARLA simulator
+
+Thanks to <https://github.com/gezp> for ROS2 Humble support for CARLA Communication.
+This ros package enables communication between Autoware and CARLA for autonomous driving simulation.
+
+# Environment
+
+| ubuntu |  ros   | carla  |    autoware     |
+| :----: | :----: | :----: | :-------------: |
+| 22.04  | humble | 0.9.15 | universe/master |
```
Comment (maxime-clem) [inline simulator/carla_autoware/README.md:12]:
```suggestion
# Supported Environment

| ubuntu |  ros   | carla  |    autoware     |
| :----: | :----: | :----: | :-------------: |
| 22.04  | humble | 0.9.15 | 2024.04 |
```

File: simulator/carla_autoware/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+# Setup
+
+## install
+
+- [Autoware.Universe](https://autowarefoundation.github.io/autoware-documentation/galactic/installation/autoware/source-installation/)
+- [CARLA Installation](https://carla.readthedocs.io/en/latest/start_quickstart/)
+- [Carla Lanelet2 Maps](https://bitbucket.org/carla-simulator/autoware-contents/src/master/maps/)
+- [Autoware launch with CARLA option](https://github.com/mraditya01/autoware_launch)
```
Comment (maxime-clem) [inline simulator/carla_autoware/README.md:21]: You should remove this line. `autoware_launch` does not need any special setup (once the PR is merged).

File: simulator/carla_autoware/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+```bash
+cd CARLA
+./CarlaUE4.sh -prefernvidia -quality-level=Low -RenderOffScreen
+```
+
+2. Run ros nodes
+
+```bash
+ros2 launch carla_autoware e2e_simulator.launch.xml map_path:=$HOME/autoware_map/carla_town_01 vehicle_model:=sample_vehicle sensor_model:=awsim_sensor_kit simulator_type:=carla
```
Comment (maxime-clem) [inline simulator/carla_autoware/README.md:47]:
```suggestion
ros2 launch autoware_launch e2e_simulator.launch.xml map_path:=$HOME/autoware_map/carla_town_01 vehicle_model:=sample_vehicle sensor_model:=awsim_sensor_kit simulator_type:=carla
```

**Review by maxime-clem — COMMENTED** [review]
I am still testing this PR but here are some comments.
Mostly I think the instructions in the README need to be improved.

File: simulator/carla_autoware/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+| ubuntu |  ros   | carla  | autoware |
+| :----: | :----: | :----: | :------: |
+| 22.04  | humble | 0.9.15 | 2024.04  |
+
+# Setup
+
+## install
+
+- [Autoware.Universe](https://autowarefoundation.github.io/autoware-documentation/galactic/installation/autoware/source-installation/)
```
Comment (maxime-clem) [inline simulator/carla_autoware/README.md:18]:
```suggestion
- [Autoware.Universe](https://autowarefoundation.github.io/autoware-documentation/main/installation/autoware/source-installation/)
```

File: simulator/carla_autoware/README.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+```
+
+3. Set initial pose (Init by GNSS)
+4. Set goal position
+5. Wait for planning
+6. Engage
+
+# Tips
+
+- If you want to edit the sensors configuration used in CARLA, edit `objects.json` located in `carla_autoware/config`.
```
Comment (maxime-clem) [inline simulator/carla_autoware/README.md:56]: Should we add something like "to match the configuration of the sensor description used by Autoware" ?

File: simulator/carla_autoware/README.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+| 22.04  | humble | 0.9.15 | 2024.04  |
+
+# Setup
+
+## install
+
+- [Autoware.Universe](https://autowarefoundation.github.io/autoware-documentation/galactic/installation/autoware/source-installation/)
+- [CARLA Installation](https://carla.readthedocs.io/en/latest/start_quickstart/)
+- [Carla Lanelet2 Maps](https://bitbucket.org/carla-simulator/autoware-contents/src/master/maps/)
+- [Python wheel for CARLA 0.9.15 Ros2 Humble communication](https://github.com/gezp/carla_ros/releases/tag/carla-0.9.15-ubuntu-22.04)
```
Comment (maxime-clem) [inline simulator/carla_autoware/README.md:21]: Can you add some extra details for this step ? I think there are 2 possible ways: install the wheel using some commands (I actually don't know how to do it); or add the egg file to the `PYTHONPATH`.

**Review by maxime-clem — CHANGES_REQUESTED** [review]

File: simulator/carla_autoware/README.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+- [CARLA Installation](https://carla.readthedocs.io/en/latest/start_quickstart/)
+- [Carla Lanelet2 Maps](https://bitbucket.org/carla-simulator/autoware-contents/src/master/maps/)
+- [Python Package for CARLA 0.9.15 Ros2 Humble communication](https://github.com/gezp/carla_ros/releases/tag/carla-0.9.15-ubuntu-22.04)
+
+  - Install the wheel using pip.
+  - OR add the egg file to the `PYTHONPATH`.
+
+  1. Download maps (y-axis inverted version) to arbitaly location
+  2. Change names and create the map folder (example: Town01) inside `autoware_map`. (`point_cloud/Town01.pcd` -> `autoware_map/Town01/pointcloud_map.pcd`, `vector_maps/lanelet2/Town01.osm`-> `autoware_map/Town01/lanelet2_map.osm`)
+  3. Create `map_projector_info.yaml` and add `projector_type: local` on the first line.
```
Comment (maxime-clem) [inline simulator/carla_autoware/README.md:27]:
```suggestion
1. Download maps (y-axis inverted version) to arbitaly location
2. Change names and create the map folder (example: Town01) inside `autoware_map`. (`point_cloud/Town01.pcd` -> `autoware_map/Town01/pointcloud_map.pcd`, `vector_maps/lanelet2/Town01.osm`-> `autoware_map/Town01/lanelet2_map.osm`)
3. Create `map_projector_info.yaml` and add `projector_type: local` on the first line.
```

**Review by maxime-clem — APPROVED** [review]

**Comment by maxime-clem** [comment]
@Kim-mins can I ask you to test this PR ? I think you are using CARLA a lot so I am very interested in your feedback.

**Comment by Kim-mins** [comment]
@maxime-clem Absolutely!

I read the instruction of the bridge, and I got some questions.

### Q1. Spawning the ego vehicle
I could build and run the latest Autoware with the modified [autoware_launch](https://github.com/autowarefoundation/autoware_launch/pull/924) and [autoware universe](https://github.com/mraditya01/autoware.universe/tree/updated_version), but I cannot know how to spawn the ego vehicle with the bridge. (seems it works differently what I've used).

First I ran Carla 0.9.15, and it was loaded with Town10. I also ran the latest Autoware with Town01(not 10), but nothing happens.
There seem some logics for [changing world](https://github.com/mraditya01/autoware.universe/blob/1f726f1c8bd64b014f8e2c2bb9f74928843918d4/simulator/carla_autoware/src/carla_autoware/carla_autoware.py#L77) and [spawning the ego vehicle](https://github.com/mraditya01/autoware.universe/blob/1f726f1c8bd64b014f8e2c2bb9f74928843918d4/simulator/carla_autoware/src/carla_autoware/carla_autoware.py#L99), but I cannot how to use it.

### Q2. Building the bridge
I also tried to build the bridge with the following commands,
```
cd colcon_ws
colcon build --symlink-install
```
but the log is as below, which seems the build was unsuccessful:
```
Starting >>> carla_autoware
--- stderr: carla_autoware                   
/usr/lib/python3/dist-packages/setuptools/command/easy_install.py:158: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
package init file 'src/carla_autoware/__init__.py' not found (or not a regular file)
---
Finished <<< carla_autoware [1.48s]
```

Actually, I could not know where to run the command. I didn't know where the `colcon_ws` is.

Thank you for all your hard work! I've been really looking forward to it.

**Comment by Kim-mins** [comment]
And this could be minor, `README.md`'s content `Supported Environment` might be changed to the Environment Setup of this PR (I found it also supports the latest version after building and running 2024.04 :smiling_face_with_tear:)

**Comment by mraditya01** [comment]
@Kim-mins Thank you very much for the Feedback! We have updated the README file to provide more detailed explanation regarding the code and parameter used.
I will also answer some of the questions,
# Q1. Spawning the Ego vehicle
Ego vehicle should have been spawned automatically at random location by default as you run the bridge. You can change the  location  to any specific coordinates (x, y, z) and orientation (roll, pitch, yaw)  by setting the `spawn_point` parameter in the `carla_autoware.launch.xml` file.
Same as the world map for CARLA, it should load automatically based on `carla_map` parameter on `carla_autoware.launch.xml`, with the default value set to `Town01`. If the map is not loaded on CARLA, there may be some error during the bridge initialization, such as timeout.

# Q2. Building the bridge
I apologize for the confusion. I neglected to update the instructions for building the bridge individually, which still referenced the old `cd colcon_ws` command, which no longer exists. The README has now been updated with the correct instructions.

# Q3. Frequency Annotation
By default, our bridge operates in synchronous mode with a 0.05-second timestep. The performance is not negatively impacted because the annotated frequency only affects the publishing rate of ROS messages, not the sensor tick rate within CARLA itself. In fact, reducing the publishing frequency can improve simulation time performance, as fewer computationally intensive functions during CARLA -> Autoware data conversion, such as `create_cloud`, are called.

# Additional Question

> And this could be minor, README.md's content Supported Environment might be changed to the Environment Setup of this PR (I found it also supports the latest version after building and running 2024.04 🥲)

Thank you for testing the latest version! I'll update the supported versions once the 2024.05 version is released so we have more stable version. Small updates can occasionally disrupt the bridge. (Happens before when Autoware required the `traffic_signal` message as input for initialization and the bridge suddenly not working 😓).

**Comment by Kim-mins** [comment]
Thank you for the response and modifications @mraditya01!

Actually, I'm not familiar to ros2. Could you please provide me more details?

Here are the commands I executed:
```
# Build bridge
colcon build --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=Release
# Run Carla
./CarlaUE4.sh -prefernvidia -quality-level=Low -RenderOffScreen
# Run Autoware
ros2 launch autoware_launch e2e_simulator.launch.xml map_path:=$HOME/autoware_map/Town01 vehicle_model:=sample_vehicle sensor_model:=awsim_sensor_kit simulator_type:=carla carla_map:=Town01
```

and I got RViz with no pose initialization, and Carla with Town10.

At this state, I guess I should run `carla_ros.py`, but I could not run it.
I ran the command `python3 -m src.carla_autoware.carla_autoware`at the path `autoware.universe/simulator/carla_autoware`, but I got the error below:

```
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/workspace/src/universe/autoware.universe/simulator/carla_autoware/src/carla_autoware/carla_autoware.py", line 21, in <module>
    from .carla_ros import carla_interface
  File "/workspace/src/universe/autoware.universe/simulator/carla_autoware/src/carla_autoware/carla_ros.py", line 20, in <module>
    from autoware_auto_control_msgs.msg import AckermannControlCommand
ModuleNotFoundError: No module named 'autoware_auto_control_msgs'
```

It seems I should install dependencies, but I have no idea to. Could you please help me?

Thank you!

**Comment by maxime-clem** [comment]
@Kim-mins 
- Did you also checkout the autoware_launch PR ?
- Did you setup the "Python Package for CARLA 0.9.15 Ros2 Humble communication" ?

If yes to the above, then can you share your logfile ? I want to make sure there is no issue with the `carla_autoware` node.
(If you want we can discuss your issues on Discord. It will be easier to help you that way)

**Comment by Kim-mins** [comment]
@maxime-clem 

I did not noticed your mention on Discord. I think we can continue the discussion there.

Thanks!

**Pull request merged**